stripe-ui-core/src/test/java/com/stripe/android/uicore/elements/TextFieldTest.kt

```diff
@@ -0,0 +1,203 @@
+import androidx.compose.ui.test.performSemanticsAction
+import androidx.compose.ui.test.performTextInput
+import androidx.compose.ui.text.TextRange
+import androidx.compose.ui.text.input.ImeAction
+import androidx.compose.ui.text.input.KeyboardCapitalization
+import androidx.compose.ui.text.input.KeyboardType
+import androidx.compose.ui.text.input.VisualTransformation
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class TextFieldTest {
+    @get:Rule
+    val composeTestRule = createComposeRule()
+
+    @Test
+    fun `On initial value passed, selection should be at the start of the field`() {
+        composeTestRule.setContent {
+            TestTextField(initialValue = "A1B")
+        }
+
+        composeTestRule
+            .onNodeWithTag(TEST_TAG)
+            .assert(hasTextSelection(TextRange(3)))
+    }
+
+    @Test
+    fun `On input, selection should move with input`() {
+        composeTestRule.setContent {
+            TestTextField(initialValue = "A1B")
+        }
+
+        val textField = composeTestRule.onNodeWithTag(TEST_TAG)
+
+        textField.performTextInput("2")
+        textField.performTextInput("C")
+
+        textField.assert(hasTextSelection(TextRange(5)))
+    }
+
+    @Test
+    fun `On input in middle of text, selection should move with input`() {
+        composeTestRule.setContent {
+            TestTextField(initialValue = "A1B")
+        }
+
+        val textField = composeTestRule.onNodeWithTag(TEST_TAG)
+
+        textField.performTextInputSelection(TextRange(1))
+
+        textField.performTextInput("2")
+        textField.performTextInput("C")
+
+        textField.assert(hasTextSelection(TextRange(3)))
+    }
+
+    @Test
+    fun `On text complete, should be able to change selection without changing text`() {
+        composeTestRule.setContent {
+            TestTextField(initialValue = null)
+        }
+
+        val textField = composeTestRule.onNodeWithTag(TEST_TAG)
+
+        textField.performTextInput("A1B2C3")
+
+        textField.performTextInputSelection(TextRange(4))
+        textField.assert(hasTextSelection(TextRange(4)))
+
+        textField.performTextInputSelection(TextRange(2))
+        textField.assert(hasTextSelection(TextRange(2)))
+
+        textField.performTextInputSelection(TextRange(1, 4))
+        textField.assert(hasTextSelection(TextRange(1, 4)))
+    }
+
+    @Test
+    fun `On text complete and attempted input, selection should not change`() {
+        composeTestRule.setContent {
+            TestTextField(initialValue = null)
+        }
+
+        val textField = composeTestRule.onNodeWithTag(TEST_TAG)
+
+        textField.performTextInput("A1B2C3")
+        textField.performTextInput("D4E")
+
+        textField.assert(hasTextSelection(TextRange(6)))
```

Comment:
probably worth including somewhere in this test that the text field's max input is 6, which is why the text is complete after the first text input